### PR TITLE
Use nova internal input Secrets

### PR DESCRIPTION
**Pull request by gibizer**
Today the end user configures a Secret by name  and a set of PasswordSelectors (i.e. key names in the Secret) on the Nova CRD to provide credentials. Then the Nova CRD controller passes this secret forward to each of the Nova subCRDs (e.g. NovaAPI, NovaScheduler). 
This has two downsides:
1. The password selector information needs to be duplicated across all the sub CRD Specs even if the content of it is exactly the same as the parent.
2. If Nova gets a big Secret that has a lot of non Nova related fields then Nova distributes this unrelated, security sensitive information further down the system and as such increasing the security bug surface. 

So this PR changes how Nova handles Secrets internally. The user still provides a Secret and a set of PasswordSelectors to Nova CRD. But the nova controller will read out only the relevant parts from that user defined Secret and creates smaller nova specific secrets for distribution to the sub CRDs. Also the internal Secrets will have hard coded key names as it is intended for machine to machine communication so flexible key names are not needed.

There will be N+1 internal Secret created by the nova controller:
* One Secret for the top level services. This always contains the cell0 and API DB credentials as well as the metadata shared secret.
* One Secret per NovaCell. This only contans the API DB credentials if the cell has API DB access. Also it only contains the metadata shared secret if the metadata service is deployed in the cell.

Both kind of Secret contains the nova service password.

**There are CRD changes**:
* The PasswordSelector fields are removed from the sub CR Specs as the internal secrets has predefined key names.
* The CellDatabasePassword field is moved from Nova.Spec.PasswordSelector to Nova.Spec.CellTemplate[].PasswordSelector to allow it being set differently per cell.
    
As the nova controller now reads the content of the Nova.Spec.Secret and not just passing it forward a watch is added to the nova controller to detect when the internal Secrets are needs to be updated. For the same reason Nova.Status.Condition now has a new InputReady condition.

After this PR we can do the following follow up PRs:
*  PR to allow that sensitive information generated by Nova CRD controller and provided to the sub CRDs to be 
moved from Spec fields to the internal Secret.
* PR use immutable Secrets for the internal Secrets and therefore remove the need in sub CRD controllers to watch for Secret changes (a possible performance bottleneck)

Related: https://issues.redhat.com/browse/OSPRH-98

### CODE REVIEW

File: controllers/nova_controller.go [hunk 1]
```diff
@@ -1361,6 +1404,99 @@ func (r *NovaReconciler) ensureCellMapped(
...
+	h *helper.Helper,
+	instance *novav1.Nova,
+	cellName string,
+	cellTemplate novav1.NovaCellTemplate,
+	externalSecret corev1.Secret,
+) (string, error) {
+
+	// NOTE(gibi): We can move other sensitive data to the internal Secret from
+	// the NovaCellSpec fields, possibly hostnames or usernames.
+	// XXX(gibi): Move the transport_url from from the MQ secret to the internal secret
```
Comment (gibizer) [inline controllers/nova_controller.go:1420]: Do you think this is a good idea?

**Comment by gibizer** [comment]
I did local testing and everything seems to work as expected.

The top level and the cell level internal secret is created (marked with bold below):
> $ oc get secret | grep nova
> **nova                                                           Opaque                                4      30m**
> nova-api-config-data                                           Opaque                                4      30m
> **nova-cell0                                                     Opaque                                3      30m**
> nova-cell0-conductor-config-data                               Opaque                                3      30m
> nova-cell0-conductor-scripts                                   Opaque                                1      30m
> nova-cell0-manage-config-data                                  Opaque                                3      30m
> nova-cell0-manage-scripts                                      Opaque                                1      30m
> **nova-cell1                                                     Opaque                                3      30m**
> nova-cell1-conductor-config-data                               Opaque                                3      30m
> nova-cell1-conductor-scripts                                   Opaque                                1      30m
> nova-cell1-manage-config-data                                  Opaque                                3      29m
> nova-cell1-manage-scripts                                      Opaque                                1      29m
> nova-cell1-novncproxy-config-data                              Opaque                                3      30m
> nova-metadata-config-data                                      Opaque                                4      30m
> nova-nova-dockercfg-w97db                                      kubernetes.io/dockercfg               1      32m
> nova-nova-token-v5pdq                                          kubernetes.io/service-account-token   4      32m
> nova-scheduler-config-data                                     Opaque                                3      30m
> rabbitmq-transport-url-nova-api-transport                      Opaque                                1      30m
> rabbitmq-transport-url-nova-cell1-transport                    Opaque                                1      30m
> 

With the proper hard coded fields
```
$ oc describe secret nova nova-cell0 nova-cell1
Name:         nova
Namespace:    openstack
Labels:       nova.openstack.org/name=nova
              nova.openstack.org/namespace=openstack
              nova.openstack.org/uid=66298817-8042-4fc2-88ec-88d53896984a
Annotations:  <none>

Type:  Opaque

Data
====
APIDatabasePassword:   8 bytes
CellDatabasePassword:  8 bytes
MetadataSecret:        10 bytes
ServicePassword:       8 bytes


Name:         nova-cell0
Namespace:    openstack
Labels:       nova.openstack.org/name=nova
              nova.openstack.org/namespace=openstack
              nova.openstack.org/uid=66298817-8042-4fc2-88ec-88d53896984a
Annotations:  <none>

Type:  Opaque

Data
====
APIDatabasePassword:   8 bytes
CellDatabasePassword:  8 bytes
ServicePassword:       8 bytes


Name:         nova-cell1
Namespace:    openstack
Labels:       nova.openstack.org/name=nova
              nova.openstack.org/namespace=openstack
              nova.openstack.org/uid=66298817-8042-4fc2-88ec-88d53896984a
Annotations:  <none>

Type:  Opaque

Data
====
APIDatabasePassword:   8 bytes
CellDatabasePassword:  8 bytes
ServicePassword:       8 bytes
```

The subCRs got the proper internal secrets:
```
$ oc get NovaAPI/nova-api NovaScheduler/nova-scheduler NovaMetadata/nova-metadata NovaConductor/nova-cell0-conductor NovaConductor/nova-cell1-conductor NovaNoVNCProxy/nova-cell1-novncproxy -o=jsonpath="{range .items[*]}{.metadata.name}{'\t'}{.spec.secret}{'\n'}{end}"
nova-api	nova
nova-scheduler	nova
nova-metadata	nova
nova-cell0-conductor	nova-cell0
nova-cell1-conductor	nova-cell1
nova-cell1-novncproxy	nova-cell1
```

**Comment by gibizer** [comment]
recheck
```
Error: Node request 200-0006159170 failed
```

**Review by mrKisaoLamb — APPROVED** [review]
Nice work!

File: controllers/nova_controller.go [hunk 2]
```diff
@@ -1361,6 +1404,99 @@ func (r *NovaReconciler) ensureCellMapped(
...
+	data := map[string]string{
+		ServicePasswordSelector:      string(externalSecret.Data[instance.Spec.PasswordSelectors.Service]),
+		CellDatabasePasswordSelector: string(externalSecret.Data[cellTemplate.PasswordSelectors.Database]),
+	}
+
+	if cellTemplate.HasAPIAccess {
+		data[APIDatabasePasswordSelector] = string(externalSecret.Data[instance.Spec.PasswordSelectors.APIDatabase])
+	}
+
+	if *cellTemplate.MetadataServiceTemplate.Replicas > 0 {
```
[the remark this answers is not in the document]
Comment (gibizer) [reply controllers/nova_controller.go:1430]: Done.

File: controllers/nova_controller.go (hunk 1 above)
Comment (gibizer) [reply controllers/nova_controller.go:1420]: I will spin up a separate PR for the move.

File: test/functional/base_test.go [hunk 3]
```diff
@@ -302,10 +302,24 @@ func CreateNovaSecret(namespace string, name string) *corev1.Secret {
...
+		map[string][]byte{
+			"NovaPassword":              []byte("service-password"),
+			"NovaAPIDatabasePassword":   []byte("api-database-password"),
+			"MetadataSecret":            []byte("metadata-secret"),
+			"NovaCell0DatabasePassword": []byte("cell0-database-password"),
+			"NovaCell1DatabasePassword": []byte("cell1-database-password"),
+			"NovaCell2DatabasePassword": []byte("cell2-database-password"),
 		},
 	)
 }
```
Comment (SeanMooney) [inline test/functional/base_test.go:325]: ok ok took me a sec to understand what this was for.
its so you can test using differnt passwords per cell.

**Review by SeanMooney — APPROVED** [review]
i think there are a few things we can adress in a follow up but i do not see anything to hold this at present.

The one thing i will say is i expect this to potentially break some fo our kuttl tests.
i think that can be adressed in a followup since we do not have them in ci. but we should add them to ci sooner rahter then later.

**Pull request merged**